nemo/lightning/pytorch/callbacks/model_checkpoint.py

```diff
@@ -320,6 +322,30 @@ def _ema_callback(self, trainer: 'pytorch_lightning.Trainer'):
                 ema_callback = callback
         return ema_callback
 
+    def _drop_optimizer_states(self, trainer, filepath: Union[str, Path], storage_options: Optional[Any]) -> None:
+        from nemo.utils.get_rank import is_global_rank_zero
+
+        # Get list of saved checkpoints
+        checkpoints = self._get_checkpoints_list(filepath)
+
+        # Drop optimizer states
+        checkpoint_index = len(checkpoints) - self.save_last_n_optim_states - 1
+        if len(checkpoints) > self.save_last_n_optim_states:
+            checkpoint_path = Path(checkpoints[checkpoint_index])
+            checkpoint_path = (
+                checkpoint_path / "weights" if os.path.isdir(checkpoint_path / "weights") else checkpoint_path
+            )
+            print(f'dropping {checkpoint_path}')
+
+            ## all odd files contain the optimizer states
+            distckpt_files = [f for f in os.listdir(checkpoint_path) if f.endswith("distcp")]
+            if is_global_rank_zero():
+                for f in distckpt_files:
+                    idx = int(f.strip(".distcp").split('_')[-1])
+                    if idx % 2 == 1:
+                        os.remove(checkpoint_path / f)
```

Comment:
The implementation of `remove_sharded_tensors` should be sth like:
1. Load metadata (see [here](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/dist_checkpointing/strategies/torch.py#L819))
2. Check if there are any files that start with that separation_hint (otherwise raise error?) and remove them
3. Remove all entries from the metadata that start with a given prefix
    1. metadata consists of 3 parts: `state_dict_metadata`, `planner_data`, `storage_data` - you should remove all entries from all parts
4. Resave the metadata
   1. See [here](https://github.com/pytorch/pytorch/blob/main/torch/distributed/checkpoint/filesystem.py#L582)

Reply:
For consistency reasons, I would suggest the following order:
1. Save tmp metadata
2. Rename orig (main) metadata to bck
3. Rename tmp metadata to orig (main)
4. Remove .distcp files
5. Remove bck orig metadata

Reply:
Thanks for the detailed suggestions. I've made some changes to this PR and the corresponding megatron one to address this, though this part I'm not super clear on:
>metadata consists of 3 parts: `state_dict_metadata`, `planner_data`, `storage_data` - you should remove all entries from all parts

Could you take a look at the megatron MR and let me know what I am missing?

Reply:
explained in MLM MR